application/src/main/java/org/togetherjava/tjbot/commands/moderation/modmail/ModMailCommand.java

```diff
@@ -0,0 +1,165 @@
+
+public final class ModMailCommand extends SlashCommandAdapter {
+
+    private static final Logger logger = LoggerFactory.getLogger(ModMailCommand.class);
+    private static final String COMMAND_NAME = "modmail";
+    private static final String OPTION_MESSAGE = "message";
+    private static final String OPTION_STAY_ANONYMOUS = "stay-anonymous";
+    private static final String OPTION_SERVER_GUILD = "server-guild";
+    private static final int COOLDOWN_DURATION_VALUE = 30;
+    private static final ChronoUnit COOLDOWN_DURATION_UNIT = ChronoUnit.MINUTES;
+    private static final Color AMBIENT_COLOR = Color.black;
+    private final Cache<Long, Instant> authorIdToLastCommandInvocation = createCooldownCache();
+    private final Predicate<String> modMailChannelNamePredicate;
+
+
+    /**
+     * Creates a new instance.
+     *
+     * @param jda the JDA instance to use to retrieve guildCache
+     * @param config the config to use for this
+     */
+    public ModMailCommand(JDA jda, Config config) {
+        super(COMMAND_NAME, "Send a message to the moderators of the selected guild",
+                CommandVisibility.GLOBAL);
+
+        OptionData messageOption = new OptionData(OptionType.STRING, OPTION_MESSAGE,
+                "What do you want to tell them?", true);
+        OptionData guildOption = new OptionData(OptionType.STRING, OPTION_SERVER_GUILD,
+                "Your server guild ID", true);
+        OptionData anonymousOption = new OptionData(OptionType.BOOLEAN, OPTION_STAY_ANONYMOUS,
+                "if set, your name is hidden", true);
+
+        List<Command.Choice> choices = jda.getGuildCache()
+            .stream()
+            .map(guild -> new Command.Choice(guild.getName(), guild.getIdLong()))
+            .toList();
+
+        guildOption.addChoices(choices);
+
+        getData().addOptions(messageOption, guildOption, anonymousOption);
+
+        modMailChannelNamePredicate =
+                Pattern.compile(config.getModMailChannelPattern()).asMatchPredicate();
+    }
+
+    private Cache<Long, Instant> createCooldownCache() {
+        return Caffeine.newBuilder()
+            .maximumSize(1_000)
+            .expireAfterAccess(COOLDOWN_DURATION_VALUE, TimeUnit.of(COOLDOWN_DURATION_UNIT))
+            .build();
+    }
+
+    @Override
+    public void onSlashCommand(SlashCommandInteractionEvent event) {
+        long userId = event.getUser().getIdLong();
+        event.deferReply().setEphemeral(true).queue();
+
+        if (isChannelOnCooldown(userId)) {
+            event.reply("Can only be used once per %s minutes.".formatted(COOLDOWN_DURATION_VALUE))
+                .setEphemeral(true)
+                .queue();
+            return;
+        }
+        authorIdToLastCommandInvocation.put(userId, Instant.now());
+
+        Optional<TextChannel> modMailAuditLog = getChannel(event);
+        if (modMailAuditLog.isEmpty()) {
+            logger.warn("Cannot find the designated modmail channel in this server.");
```

Comment:
that message isnt very helpful to find the issue in hindsight. put urself into the shoes of the guy who has to read the logs on a server. they know that something went wrong, but dont know anything about why it went wrong.

the message needs:
* which server id?
* whats the pattern for that guild?

Reply:
Pattern of the guild? I can get the pattern for a channel .

Al though I can do this  `Objects.requireNonNull(event.getGuild()).getName())` but not sure what you think of it?

Reply:
requireNonNull won't help this if this event is not from a guild

Reply:
Alright, but should I still get the guild name this way

Reply:
> Pattern of the guild? I can get the pattern for a channel .
i meant whatever that `getChannel()` method uses to locate it. i assumed it goes by channel pattern.

all im saying is that the logger message isnt helpful as u made it. it needs to tell someone all the available info. just put urself into the shoes of someone seeing ur logs in the log file. what infos do u need to figure out why it errored? thats the info u have to provide.

Reply:
So you mean like this?

```java
        Optional<TextChannel> modMailAuditLog = getModMailChannel(event.getJDA(), userGuildId);
        if (modMailAuditLog.isEmpty()) {
            logger.warn(
                    "Cannot find the designated modmail channel in server %s server by the id of %d ."
                        .formatted(getModMailChannel(event.getJDA(), userGuildId), userGuildId));
            return;
        }
```